File: src/test/java/org/jetbrains/plugins/ideavim/helper/SearchHelperTest.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2022 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package org.jetbrains.plugins.ideavim.helper;

import com.maddyhome.idea.vim.api.VimSearchHelperBase;
import com.maddyhome.idea.vim.command.VimStateMachine;
import com.maddyhome.idea.vim.helper.SearchHelperKtKt;
import org.jetbrains.plugins.ideavim.SkipNeovimReason;
import org.jetbrains.plugins.ideavim.TestWithoutNeovim;
import org.jetbrains.plugins.ideavim.VimTestCase;

public class SearchHelperTest extends VimTestCase {
  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindNextWord() {
    String text = "first second";
    int nextWordPosition = (int)VimSearchHelperBase.Companion.findNextWord(text, 0, text.length(), 1, true, false);

    assertEquals(nextWordPosition, text.indexOf("second"));
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindSecondNextWord() {
    String text = "first second third";
    int nextWordPosition = (int)VimSearchHelperBase.Companion.findNextWord(text, 0, text.length(), 2, true, false);

    assertEquals(nextWordPosition, text.indexOf("third"));
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindAfterLastWord() {
    String text = "first second";
    int nextWordPosition = (int)VimSearchHelperBase.Companion.findNextWord(text, 0, text.length(), 3, true, false);

    assertEquals(nextWordPosition, text.length());
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindPreviousWord() {
    String text = "first second";
    int previousWordPosition =
      (int)VimSearchHelperBase.Companion.findNextWord(text, text.indexOf("second"), text.length(), -1, true, false);

    //noinspection ConstantConditions
    assertEquals(previousWordPosition, text.indexOf("first"));
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindSecondPreviousWord() {
    String text = "first second third";
    int previousWordPosition =
      (int)VimSearchHelperBase.Companion.findNextWord(text, text.indexOf("third"), text.length(), -2, true, false);

    //noinspection ConstantConditions
    assertEquals(previousWordPosition, text.indexOf("first"));
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindBeforeFirstWord() {
    String text = "first second";
    int previousWordPosition =
      (int)VimSearchHelperBase.Companion.findNextWord(text, text.indexOf("second"), text.length(), -3, true, false);

    //noinspection ConstantConditions
    assertEquals(previousWordPosition, text.indexOf("first"));
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testFindPreviousWordWhenCursorOutOfBound() {
    String text = "first second";
    int previousWordPosition =
      (int)VimSearchHelperBase.Companion.findNextWord(text, text.length(), text.length(), -1, true, false);

    assertEquals(previousWordPosition, text.indexOf("second"));
  }

  public void testMotionOuterWordAction() {
    doTest("va(", "((int) nu<caret>m)", "<selection>((int) num)</selection>", VimStateMachine.Mode.VISUAL,
           VimStateMachine.SubMode.VISUAL_CHARACTER);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringInsideDoubleQuotes() {
    String text = "abc\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringWithoutClosingDoubleQuote() {
    String text = "abcdef\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOnUnpairedSingleQuote() {
    String text = "abc\"d'ef\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOutsideOfDoubleQuotesPair() {
    String text = "abc\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 2, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringEscapedDoubleQuote() {
    String text = "abc\\\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOddNumberOfDoubleQuotes() {
    String text = "abc\"def\"gh\"i";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringInsideSingleQuotesPair() {
    String text = "abc\"d'e'f\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 6, false);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOnOpeningDoubleQuote() {
    String text = "abc\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 3, true);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOnClosingDoubleQuote() {
    String text = "abc\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 7, true);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringWithoutQuotes() {
    String text = "abcdefghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringDoubleQuoteInsideSingleQuotes() {
    String text = "abc'\"'ef\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringSingleQuotesAreTooFarFromEachOtherToMakePair() {
    String text = "abc'\"de'f\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 5, true);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringDoubleQuoteInsideSingleQuotesIsInsideSingleQuotedString() {
    String text = "abc'\"'def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 4, false);
    assertTrue(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringAfterClosingDoubleQuote() {
    String text = "abc\"def\"ghi";
    boolean inString = SearchHelperKtKt.checkInString(text, 9, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringOnMiddleDoubleQuote() {
    String text = "abc\"def\"gh\"i";
    boolean inString = SearchHelperKtKt.checkInString(text, 7, true);
    assertFalse(inString);
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  public void testCheckInStringBetweenPairs() {
    String text = "abc\"def\"gh\"ij\"k";
    boolean inString = SearchHelperKtKt.checkInString(text, 8, true);
    assertFalse(inString);
  }
}

File: src/test/java/org/jetbrains/plugins/ideavim/helper/SearchHelperTest.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package org.jetbrains.plugins.ideavim.helper

import com.intellij.openapi.application.ApplicationManager
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.group.findBlockRange
import com.maddyhome.idea.vim.newapi.vim
import com.maddyhome.idea.vim.state.mode.Mode
import com.maddyhome.idea.vim.state.mode.SelectionType
import org.jetbrains.plugins.ideavim.SkipNeovimReason
import org.jetbrains.plugins.ideavim.TestWithoutNeovim
import org.jetbrains.plugins.ideavim.VimTestCase
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

@Suppress("SpellCheckingInspection")
class SearchHelperTest : VimTestCase() {
  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindNextWord() {
    val text = "first second"
    configureByText(text)
    val nextWordPosition = injector.searchHelper.findNextWord(fixture.editor.vim, 0, 1, bigWord = true)
    kotlin.test.assertEquals(nextWordPosition, text.indexOf("second"))
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindSecondNextWord() {
    val text = "first second third"
    configureByText(text)
    val nextWordPosition = injector.searchHelper.findNextWord(fixture.editor.vim, 0, 2, bigWord = true)
    kotlin.test.assertEquals(nextWordPosition, text.indexOf("third"))
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindAfterLastWord() {
    val text = "first second"
    configureByText(text)
    val nextWordPosition = injector.searchHelper.findNextWord(fixture.editor.vim, 0, 3, bigWord = true)
    kotlin.test.assertEquals(nextWordPosition, text.length)
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindPreviousWord() {
    val text = "first second"
    configureByText(text)
    val previousWordPosition =
      injector.searchHelper.findNextWord(fixture.editor.vim, text.indexOf("second"), -1, bigWord = true)
    kotlin.test.assertEquals(previousWordPosition, text.indexOf("first"))
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindSecondPreviousWord() {
    val text = "first second third"
    configureByText(text)
    val previousWordPosition =
      injector.searchHelper.findNextWord(fixture.editor.vim, text.indexOf("third"), -2, bigWord = true)
    kotlin.test.assertEquals(previousWordPosition, text.indexOf("first"))
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindBeforeFirstWord() {
    val text = "first second"
    configureByText(text)
    val previousWordPosition =
      injector.searchHelper.findNextWord(fixture.editor.vim, text.indexOf("second"), -3, bigWord = true)
    kotlin.test.assertEquals(previousWordPosition, text.indexOf("first"))
  }

  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindPreviousWordWhenCursorOutOfBound() {
    val text = "first second"
    configureByText(text)
    val previousWordPosition = injector.searchHelper.findNextWord(fixture.editor.vim, text.length, -1, bigWord = true)
    kotlin.test.assertEquals(previousWordPosition, text.indexOf("second"))
  }


  @TestWithoutNeovim(reason = SkipNeovimReason.NOT_VIM_TESTING)
  @Test
  fun testFindAllIgnoreCaseOverwritesSmartCase() {
    val text = "Lorem ipsum lorem ipsum"
    configureByText(text)

    val capitalMatch = TextRange(0, 5)
    val lowerMatch = TextRange(12, 17)

    val search = { ignoreCase: Boolean ->
      injector.searchHelper.findAll(fixture.editor.vim, "\\<Lorem\\>", 0, -1, ignoreCase)
    }

    // Ensure ignore and smart case are off
    enterCommand("set noignorecase")
    enterCommand("set nosmartcase")

    kotlin.test.assertEquals(listOf(capitalMatch), search(false))
    // Even if both are off, ignore case should still ignore cases
    kotlin.test.assertEquals(listOf(capitalMatch, lowerMatch), search(true))

    enterCommand("set smartcase")
    kotlin.test.assertEquals(listOf(capitalMatch), search(false))
    kotlin.test.assertEquals(listOf(capitalMatch, lowerMatch), search(true))
  }

  @Test
  fun testMotionOuterWordAction() {
    doTest(
      "va(",
      "((int) nu<caret>m)",
      "<selection>((int) num)</selection>",
      Mode.VISUAL(SelectionType.CHARACTER_WISE),
    )
  }

  @ParameterizedTest
  @MethodSource("findBlockRangeTestCases")
  fun findBlockRange(testCase: FindBlockRangeTestCase) {
    val (_, text, type, count, isOuter, expected) = (testCase)
    configureByText(text)
    ApplicationManager.getApplication().runReadAction {
      val actual = findBlockRange(fixture.editor.vim, fixture.editor.vim.currentCaret(), type, count, isOuter)
      kotlin.test.assertEquals(expected, actual)
    }
  }

  companion object {
    data class FindBlockRangeTestCase(
      val description: String,
      val text: String,
      val type: Char,
      val count: Int,
      val isOuter: Boolean,
      val expected: TextRange?,
    )

    @JvmStatic
    fun findBlockRangeTestCases(): Stream<FindBlockRangeTestCase> {
      return Stream.of(
        FindBlockRangeTestCase("no inner match when range selection", "${s}aa${se}", '(', 1, isOuter = false, expected = null),
        FindBlockRangeTestCase("no inner match when no range selection", "${s}a${se}", '(', 1, isOuter = false, expected = null),
        FindBlockRangeTestCase("inner match without range selection", "(${c}a)", '(', 1, isOuter = false, expected = TextRange(1, 2)),
        FindBlockRangeTestCase("outer match without range selection", "(${c}a)", '(', 1, isOuter = true, expected = TextRange(0, 3)),
        FindBlockRangeTestCase("inner match with range selection", "(${c}${s}a${se})", '(', 1, isOuter = false, expected = TextRange(1, 2)),
        FindBlockRangeTestCase("outer match with range selection", "(${c}${s}a${se})", '(', 1, isOuter = true, expected = TextRange(0, 3)),
        FindBlockRangeTestCase("caret at begin inner match forward", "(a)${c}(a)", '(', 1, isOuter = false, expected = TextRange(4, 5)),
        FindBlockRangeTestCase("caret at begin outer match forward", "(a)${c}(a)", '(', 1, isOuter = true, expected = TextRange(3, 6)),
        FindBlockRangeTestCase("caret at end inner match backward", "(a${c})(a)", '(', 1, isOuter = false, expected = TextRange(1, 2)),
        FindBlockRangeTestCase("caret at end outer match backward", "(a${c})(a)", '(', 1, isOuter = true, expected = TextRange(0, 3)),
        FindBlockRangeTestCase("inner match first paren", "( (${c}a) )", '(', 1, isOuter = false, expected = TextRange(3, 4)),
        FindBlockRangeTestCase("outer match first paren", "( (${c}a) )", '(', 1, isOuter = true, expected = TextRange(2, 5)),
        FindBlockRangeTestCase("caret at begin inner match first forward", "${c}( (a) )", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("caret at begin outer match first forward", "${c}( (a) )", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match multiple occurence", "( (${c}a) )", '(', 2, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match multiple occurence", "( (${c}a) )", '(', 2, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("expand inner selection to first occurence", "(${c}${s}a${se}a)", '(', 1, isOuter = false, expected = TextRange(1, 3)),
        FindBlockRangeTestCase("expand outer selection to first occurence", "(${c}${s}a${se}a)", '(', 1, isOuter = true, expected = TextRange(0, 4)),
        FindBlockRangeTestCase("expand inner selection to next occurence", "( (${c}${s}aa${se}) )", '(', 1, isOuter = false, expected = TextRange(1, 7)),
        FindBlockRangeTestCase("expand outer selection to next occurence", "( ${c}${s}(aa)${se} )", '(', 1, isOuter = true, expected = TextRange(0, 8)),
        FindBlockRangeTestCase("paren not in string inner match", "( (\"${c}a\") )", '(', 1, isOuter = false, expected = TextRange(3, 6)),
        FindBlockRangeTestCase("paren not in string outer match", "( (\"${c}a\") )", '(', 1, isOuter = true, expected = TextRange(2, 7)),
        FindBlockRangeTestCase("start end paren in string inner match", "( \"(a${c}a)\" )", '(', 1, isOuter = false, expected = TextRange(4, 6)),
        FindBlockRangeTestCase("start end paren in string outer match", "( \"(a${c}a)\" )", '(', 1, isOuter = true, expected = TextRange(3, 7)),
        FindBlockRangeTestCase("only start paren in string inner match not match paren outside", "\"(a${c}a\")", '(', 1, isOuter = false, expected = null),
        FindBlockRangeTestCase("only start paren in string outer match not match paren outside", "(\"a${c}a)\"", '(', 1, isOuter = true, expected = null),
        FindBlockRangeTestCase("inner match exclude start paren in string when caret at start of quote", "(${c}\"(aa\")", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match exclude start paren in string when caret at start of quote", "(${c}\"(aa\")", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match exclude start paren in string when caret at end of quote", "(\"(aa${c}\")", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match exclude start paren in string when caret at end of quote", "(\"(aa${c}\")", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match not exclude start paren in string when caret in between quote", "(\"(a${c}a\")", '(', 1, isOuter = false, expected = TextRange(1, 6)), // Vim behavior differs, but we have some PSI magic and can resolve such cases
        FindBlockRangeTestCase("outer match not exclude start paren in string when caret in between quote", "(\"(a${c}a\")", '(', 1, isOuter = true, expected = TextRange(0, 7)), // Vim behavior differs, but we have some PSI magic and can resolve such cases
        FindBlockRangeTestCase("inner match exclude end paren in string when caret at start of quote", "(${c}\"aa)\")", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match exclude end paren in string when caret at start of quote", "(${c}\"aa)\")", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match exclude end paren in string when caret at end of quote", "(\"aa)${c}\")", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match exclude end paren in string when caret at end of quote", "(\"aa)${c}\")", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match exclude end paren in string when caret in between quote", "(\"a${c}a)\")", '(', 1, isOuter = false, expected = TextRange(1, 6)),
        FindBlockRangeTestCase("outer match exclude end paren in string when caret in between quote", "(\"a${c}a)\")", '(', 1, isOuter = true, expected = TextRange(0, 7)),
        FindBlockRangeTestCase("inner match exclude first line break after start paren", "(\n${c}a)", '(', 1, isOuter = false, expected = TextRange(2, 3)),
        FindBlockRangeTestCase("inner match exclude only first line break after start paren", "(\n\n${c}a)", '(', 1, isOuter = false, expected = TextRange(2, 4)),
        FindBlockRangeTestCase("inner match exclude last line break before end paren", "(${c}a\n)", '(', 1, isOuter = false, expected = TextRange(1, 2)),
        FindBlockRangeTestCase("inner match exclude only last line break before end paren", "(${c}a\n\n)", '(', 1, isOuter = false, expected = TextRange(1, 3)),
        FindBlockRangeTestCase("inner match exclude last line break and whitespace character before end paren", "(${c}a\n   )", '(', 1, isOuter = false, expected = TextRange(1, 2)),
        FindBlockRangeTestCase("inner match expand to next occurence when start end paren has line break", "(a(\n${c}${s}a${se}\n)c)", '(', 1, isOuter = false, expected = TextRange(1, 8)),
        FindBlockRangeTestCase("inner match expand to next occurence when end paren has line break and blank character", "(a(\n${c}${s}a${se}\n )c)", '(', 1, isOuter = false, expected = TextRange(1, 9)),
        FindBlockRangeTestCase("inner match not expand to next occurence when selection not last line", "(a(\n${c}${s}a${se}\n\n )c)", '(', 1, isOuter = false, expected = TextRange(4, 6)),
      )
    }
  }
}
